src/mpc/cli/mpc_phase2_begin.cpp

```diff
@@ -0,0 +1,116 @@
+#include "mpc_common.hpp"
+#include "snarks/groth16/mpc/mpc_utils.hpp"
+#include "snarks/groth16/mpc/phase2.hpp"
+
+#include <boost/program_options.hpp>
+#include <fstream>
+
+using namespace libzeth;
+namespace po = boost::program_options;
+
+namespace
+{
+
+// Usage:
+//   $0 phase2-begin [<options>] <linear_combination_file>
+//
+// Options:
+//   --out <file>    Initial challenge output file (mpc-challenge.bin)
+class mpc_phase2_begin : public subcommand
+{
+private:
+    std::string lin_comb_file;
+    std::string out_file;
+
+public:
+    mpc_phase2_begin() : subcommand("phase2-begin"), lin_comb_file(), out_file()
+    {
+    }
+
+private:
+    void initialize_suboptions(
+        boost::program_options::options_description &options,
+        boost::program_options::options_description &all_options,
+        boost::program_options::positional_options_description &pos) override
+    {
+        options.add_options()(
+            "out,o",
+            po::value<std::string>(),
+            "Initial challenge output file (mpc-challenge.bin)");
+        all_options.add(options).add_options()(
+            "linear_combination_file",
+            po::value<std::string>(),
+            "linear combination file");
+        pos.add("linear_combination_file", 1);
+    }
+
+    void parse_suboptions(
+        const boost::program_options::variables_map &vm) override
+    {
+        if (0 == vm.count("linear_combination_file")) {
+            throw po::error("linear_combination_file not specified");
+        }
+        lin_comb_file = vm["linear_combination_file"].as<std::string>();
+        out_file = vm.count("out") ? vm["out"].as<std::string>()
+                                   : trusted_setup_file("mpc-challenge.bin");
+    }
+
+    void subcommand_usage() override
+    {
+        std::cout << "Usage:\n  " << subcommand_name
+                  << " [<options>] <linear_combination_file>\n\n";
+    }
+
+    int execute_subcommand() override
+    {
+        if (verbose) {
+            std::cout << "lin_comb_file: " << lin_comb_file << "\n";
+            std::cout << "out: " << out_file << std::endl;
+        }
+
+        libff::enter_block("Load linear combination file");
+        srs_mpc_layer_L1<ppT> lin_comb = [&]() {
+            std::ifstream in(
+                lin_comb_file, std::ios_base::binary | std::ios_base::in);
+            return srs_mpc_layer_L1<ppT>::read(in);
+        }();
+        libff::leave_block("Load linear combination file");
+
+        // Compute circuit
+        libff::enter_block("Computing num inputs");
+        const size_t num_inputs = [this]() {
+            libsnark::protoboard<FieldT> pb;
+            init_protoboard(pb);
+            return pb.num_inputs();
+        }();
+        libff::print_indent();
+        std::cout << std::to_string(num_inputs) << std::endl;
+        libff::leave_block("Computing num inputs");
+
+        // TODO: use the hash of the linear combination as the inital transcript
```

Comment:
What's the state on that  @dtebbs?

Reply:
I recall mentioning this, but I do not recall any decision, so it's just left as a comment.  I'll make the change.  We will need to expose something for clients, so they can confirm the phase2 transcript begins with the expected digest.

Reply:
After discussing, I've added the following to this PR:
- added a `cs_hash` field to the phase2 accumulators
- this is set to the digest of the linear-combination when the initial challenge is created
- `verify_contribution` now checks that the cs_hash in response and challenge match
- `cs_hash` is used to "seed" the value of the transcript digest

Reply:
The TODO should have been removed in the latest version.